mlc_llm/transform/reorder_transform_func.py

```diff
@@ -0,0 +1,219 @@
+
+
+def analyze_func(
+    func: relax.Function,
+    pidx2binname: Dict[int, str],
+) -> Tuple[
+    List[relax.Binding],
+    Dict[relax.Var, List[relax.Binding]],
+    Dict[relax.Binding, int],
+]:
+    """Binding grouping analysis function.
+    It takes the function to be analyzed, and mapping from each raw tensor index
+    to the name of the binary file where it resides.
+
+    This analysis function
+    * computes a new order of weight fetching bindings (the bindings in form
+    `lv = params[idx]`) based on weight location on disk.
+    * collects the dataflow def-use information of the given function for
+    topological sort (particularly, it collects the consumers of each binding
+    variables and the number of variables each binding depends on).
+
+    Parameters
+    ----------
+    func : relax.Function
+        The weight transform function to be analyzed.
+
+    pidx2binname : Dict[int, str]
+        The mapping from each raw tensor index to the name of the binary
+        file where it resides.
+
+    Returns
+    -------
+    get_param_bindings : List[relax.Binding]
+        The weight fetching bindings (`lv = params[idx]`) in the new order.
+
+    var_users : Dict[relax.Var, List[relax.Binding]]
+        The consumer bindings of each binding variable.
+        Used for topological sort.
+
+    num_depending_vars : Dict[relax.Binding, int]
+        The number of variables each binding depends on.
+        Used for topological sort.
+    """
+
+    # The mapping of the weight fetching bindings in each binary file.
+    # Here empty string means the weight is not in any binary file (e.g., cached
+    # sin and cos values for rotary embeddings).
+    binname2get_param_bindings: Dict[str, List[relax.Binding]] = {"": []}
+    # The set of binding variables.
+    binding_var_set: Set[relax.Var] = set()
+    var_users: Dict[relax.Var, List[relax.Binding]] = {}
+    num_depending_vars: Dict[relax.Binding, int] = {}
+
+    # Sanity check on the function pattern.
+    assert len(func.params) == 1
+    assert isinstance(func.body, relax.SeqExpr)
+    assert len(func.body.blocks) == 1
+    assert isinstance(func.body.blocks[0], relax.DataflowBlock)
+    assert func.body.blocks[0].bindings[-1].var.same_as(func.body.body)
+
+    params = func.params[0]
+    bindings = func.body.blocks[0].bindings
+
+    # Go through each binding except the last one. (The last one is the output
+    # binding `gv = (lv, lv1, ...)`) which we ignore for analysis.
+    for binding in bindings[:-1]:
+        value = binding.value
+        binding_var_set.add(binding.var)
+        var_users[binding.var] = []
+
+        if isinstance(value, relax.TupleGetItem) and value.tuple_value.same_as(params):
+            # For weight fetching bindings (`lv = params[idx]`), we group them
+            # according to the binary file name.
+            pidx = value.index
+            if pidx not in pidx2binname:
+                binname2get_param_bindings[""].append(binding)
```

Comment:
What does this mean?

Reply:
https://github.com/mlc-ai/mlc-llm/blob/cdb78cc58e4ce61e2be15e85f1b8856ae249a3e2/mlc_llm/transform/reorder_transform_func.py#L80-L81